README
XMLGEN -- A module to generate XML and/or HTML by writing Tcl code which
          translates itself.
          (C) 2002 Harald Kirsch

WHAT IS IT?
===========

You declare the XML tags you want as Tcl commands. Then you can use
the natural bracing structure of Tcl to write XML while having all the
niceties of the script language to define shortcuts or add dynamic
content. 

A simple example based on htmlgen, a module which predefines all the
HTML tags.

set Title "A Simple Example"
html ! {
  head ! {
    title - $Title
  }
  body ! {
    h1 - $Title
    p - As you can see, title and major headline agree.
  }
}

INSTALLATION
============
This package contains three pieces to be installed in different
directories of your choice:

  1) the Tcl package xmlgen (contains htmlgen)
  2) HTML documentation for both
  3) cgi-enabled documentation, mainly here as a proof of concept and
     as an example

The QUICK INSTALL goes like

  tclsh bras-sa prefix=/usr/local cgi_prefix=/home/html/cgi install

This will create the following directories with some content:

  /usr/local/lib/xmlgen-1.0       (possibly different version number)
  /usr/local/doc/xmlgen-1.0       (dto.)
  /home/html/cgi/xmlgen-1.0	  (dto.)
    
NOTES:

1) For the cgi-enabled documentation to work you need two things:
  a) Convince your web server to run $CGDIR/index as a Tcl script.
  b) Make sure this script can find a recent tcllib. It needs the
     package ncgi.

2) If you want to set the subdirectory names yourself, then specify
*all* of PKGDIR, DOCDIR and CGIDIR instead of prefix and cgi_prefix.

3) In any case, the parent directory of PKGDIR, /usr/local/lib in the
example, MUST appear in the auto_path variable of the tclsh you intend
to use. (HINT: auto_path is initialized also from the TCLLIBPATH
environment variable.)

4) bras-sa is a standalone version of bras, the make replacement
written in Tcl to be found at http://bras.berlios.de/ . It uses the
brasfile in this directory to do its work.


LICENSE 
=======
BSD derived, see file LICENSE	  


CONTACT
=======
Harald Kirsch
pifpafuf bei gmx punkt de

----------------------
last modified: $Date: 2002/05/27 21:50:52 $


CHANGES
** This change log was extracted from `cvs log',
** sorted by date and written so that identical descriptions
** are shown only once.
**                          digestLog
**                          (C) Harald Kirsch (pifpafpuf@gmx.de)
**                          $Revision: 1.7 $
**                          $Date: 2002/02/11 18:47:30 $
-----------------------------------------------------------

1) The empty string is recognized as an attribute value pair and then
ignored to allow for optional autogenerated attribute value pairs.
2) By default, the closing tag generated by control character '-' is
now put right after the content instead of indented on the next line.
3) Formatting of tags --- in particular indentation --- is now fully
configurable.

applies to:
/cvsroot/tclxml/xmlgen/doc/xmlgen, 1.4, 2002/09/21 14:55:58;
/cvsroot/tclxml/xmlgen/doc/common.tcl, 1.4, 2002/09/21 14:55:58;
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.9, 2002/09/21 14:55:55;
/cvsroot/tclxml/xmlgen/pkgIndex.tcl, 1.7, 2002/09/21 14:55:55;
/cvsroot/tclxml/xmlgen/ANNOUNCE, 1.5, 2002/09/21 14:55:55;
-----------------------------------------------------------

Removed mostly useless test code from end of file.

applies to:
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.8, 2002/09/21 11:37:02;
-----------------------------------------------------------

This wants to become 1.3 .

applies to:
/cvsroot/tclxml/xmlgen/pkgIndex.tcl, 1.6, 2002/08/07 17:36:19;
-----------------------------------------------------------

o [puts] is no longer tweaked in [buffer] context, use [put] instead.
o [channel] works similar to [buffer]. It redirects to an output
  channel.

applies to:
/cvsroot/tclxml/xmlgen/doc/xmlgen, 1.3, 2002/08/07 17:33:14;
/cvsroot/tclxml/xmlgen/doc/sidenav, 1.2, 2002/08/07 17:33:14;
/cvsroot/tclxml/xmlgen/doc/index, 1.2, 2002/08/07 17:33:14;
/cvsroot/tclxml/xmlgen/doc/common.tcl, 1.3, 2002/08/07 17:33:13;
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.7, 2002/08/07 17:33:08;
/cvsroot/tclxml/xmlgen/pkgIndex.tcl, 1.5, 2002/08/07 17:33:08;
/cvsroot/tclxml/xmlgen/htmlgen.tcl, 1.5, 2002/08/07 17:33:08;
/cvsroot/tclxml/xmlgen/ANNOUNCE, 1.4, 2002/08/07 17:33:08;
/cvsroot/tclxml/xmlgen/sidenav.tcl, 1.3, 2002/08/07 17:33:08;
-----------------------------------------------------------

In progress...

applies to:
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.6, 2002/07/11 05:32:03;
/cvsroot/tclxml/xmlgen/webutils.tcl, 1.4, 2002/07/11 05:32:03;
/cvsroot/tclxml/xmlgen/test.tml, 1.4, 2002/07/11 05:32:03;
-----------------------------------------------------------

Working on making TML Wiki-like

applies to:
/cvsroot/tclxml/xmlgen/test.tml, 1.3, 2002/06/25 05:25:49;
-----------------------------------------------------------

Undo some formatting changes I made to Harald's code without good reason, and make my formatting conform better to his (Mea Culpa!)

applies to:
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.5, 2002/06/25 04:23:35;
/cvsroot/tclxml/xmlgen/webutils.tcl, 1.3, 2002/06/25 04:23:35;
/cvsroot/tclxml/xmlgen/test.tml, 1.2, 2002/06/25 04:23:35;
-----------------------------------------------------------

Adding the example file test.tml to the CVS repository

applies to:
/cvsroot/tclxml/xmlgen/test.tml, 1.1, 2002/06/20 14:54:44;
-----------------------------------------------------------

Some work on webutils and cleanup responsive to discussion with HK

applies to:
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.4, 2002/06/20 05:19:00;
/cvsroot/tclxml/xmlgen/htmlgen.tcl, 1.4, 2002/06/20 05:19:00;
/cvsroot/tclxml/xmlgen/webutils.tcl, 1.2, 2002/06/20 05:19:00;
/cvsroot/tclxml/xmlgen/webgen.tcl, 1.2, 2002/06/20 05:19:00;
-----------------------------------------------------------

PARSE extension allow easy parsing of options and variables for procs

applies to:
/cvsroot/tclxml/xmlgen/parse.tcl, 1.1, 2002/06/10 03:22:08;
-----------------------------------------------------------

Cleanup to make the code easier to work with and more consistently usable. TCL's 'puts' is now left alone, and new proc 'putx' performs 'put XML'. Initial work on WEBUTILS, for parsing plain text into proto-html. Fixed obvious typo in LICENSE.

applies to:
/cvsroot/tclxml/xmlgen/doc/common.tcl, 1.2, 2002/06/10 03:17:44;
/cvsroot/tclxml/xmlgen/pkgIndex.tcl, 1.4, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/xmlgen.tcl, 1.3, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/htmlgen.tcl, 1.3, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/tab.tcl, 1.2, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/sidenav.tcl, 1.2, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/LICENSE, 1.2, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/webutils.tcl, 1.1, 2002/06/10 03:17:42;
/cvsroot/tclxml/xmlgen/webgen.tcl, 1.1, 2002/06/10 03:17:42;
-----------------------------------------------------------

Fixed some typos in the docs.
Switched to using a standalone version of bras as installation driver.
This wants to be 1.2.1.

applies to:
/cvsroot/tclxml/xmlgen/doc/xmlgen, 1.2, 2002/05/27 21:50:54;
/cvsroot/tclxml/xmlgen/doc/overview, 1.2, 2002/05/27 21:50:54;
/cvsroot/tclxml/xmlgen/doc/htmlgen, 1.2, 2002/05/27 21:50:54;
/cvsroot/tclxml/xmlgen/Brasfile, 1.4, 2002/05/27 21:50:52;
/cvsroot/tclxml/xmlgen/pkgIndex.tcl, 1.3, 2002/05/27 21:50:52;
/cvsroot/tclxml/xmlgen/ANNOUNCE, 1.3, 2002/05/27 21:50:52;
/cvsroot/tclxml/xmlgen/brasfile, 1.3, 2002/05/27 21:50:52;
/cvsroot/tclxml/xmlgen/README, 1.3, 2002/05/27 21:50:52;
-----------------------------------------------------------

No longer needed, using bras now.

applies to:
/cvsroot/tclxml/xmlgen/Attic/install, 1.4, 2002/05/27 21:06:03;
-----------------------------------------------------------

Initial checkin. Will be used as general built and install
driver. (What else could be used if you don't want to use make?)

applies to:
/cvsroot/tclxml/xmlgen/bras-sa, 1.1, 2002/05/27 20:13:29;
-----------------------------------------------------------

Initial checkin of xmlgenSite.

applies to:
/cvsroot/tclxml/xmlgen/doc/xmlgenSite, 1.1, 2002/05/27 20:11:52;

parse.tcl
########################################################################
#
#  parse -- extension for parsing procedure arguments
# 
# $Revision: 1.1 $, $Date: 2002/06/10 03:22:08 $
########################################################################
set license {
Copyright (c) 2002, Edwin A. Suominen
Registered Patent Agent * http://eepatents.com
Author, PRIVARIA Secure Networking Suite
Freely available at http://www.privaria.org

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of the author nor any contributor may be used to
endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
}

package require cmdline

namespace eval ::parse {

namespace export args

#<<<<<<<<<<<< PARSE::ARGS >>>>>>>>>>>>>>#
proc args { optsList varsList args } {

### List Pre-processing
# Determine whether args is only a single variable (list?)
set singleVarFlag 1
if {    [llength $args] == 0    ||
        [regexp {^\{*(\-|\}$)} $args]  } {
    # Not possible to be a single-variable arg list
    # if there's any option present
    set singleVarFlag 0

} else {
    foreach {i} $args {
        if { [llength $i] > 1 } {
            set singleVarFlag 0
            break
        }
    }
}
# Now work on each list in turn
foreach {i} {optsList varsList args} {
    # Massage selected list if multiply escaped
    while { [llength [set $i]] == 1 &&
            [regexp {\x20} [set $i]] } {
        set $i [lindex [set $i] 0]
    }

    # Extract any default values from opts, vars lists (to be overwritten
    # for defined options)
    if { [regexp List $i] } {
        set count 0
        set x [set $i]  ;# For loop, create temp list w/ selected list contents
        foreach {j} $x {
            if { [llength $j] > 1 } {
                regexp -nocase {^([a-z0-9]+)} [lindex $j 0] null varName
                set $i [lreplace [set $i] $count $count [lindex $j 0] ]
                set xxx [lindex $j end]
                uplevel [list set $varName $xxx]
            }
            incr count
        }
    }
}
# Assign variables based on lists
if {$singleVarFlag} {
    if { [llength $varsList] > 0 } {
        uplevel [subst { set [lindex $varsList 0] {$args} } ]
    }

} else {
    uplevel [subst {
        set parseargsList [list $args]
        while { \[ ::cmdline::getopt parseargsList {$optsList} x y \] == 1 } {
            set \$x \$y }
        foreach {i} {$varsList} {j} \$parseargsList {
            # The IF corrects a bug that was very hard to spot. Can you?
            if { \[llength \$j \] > 0 } { 
                set \$i \$j
            } elseif { !\[info exists \$i \] } {
                set \$i \[list \]   ;# List w/ no args returns empty list
            }
        }
        unset parseargsList
    }]
}

### END PARSEARGS
return
}

### END PARSE.TCL
}

webgen.tcl
########################################################################
#
#  webgen -- generate web site from easily maintained source files
# 
# $Revision: 1.2 $, $Date: 2002/06/20 05:19:00 $
########################################################################
set license {
Copyright (c) 2002, Edwin A. Suominen
Registered Patent Agent * http://eepatents.com
Author, PRIVARIA Secure Networking Suite
Freely available at http://www.privaria.org

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of the author nor any contributor may be used to
endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
}

foreach {i} {webutils htmlgen} { package require $i }
namespace import -force ::htmlgen::*

# This script is called from the command line as follows:
# tclsh|wish masterfile.tml
if { $argc >= 1 } {
    set masterFileSpec [lindex $argv 0]
} else {
    set masterFileSpec masterfile.tml
}



webutils.tcl
########################################################################
#  webutils -- Toolbox of utility procs for easy web site generation
# $Revision: 1.4 $, $Date: 2002/07/11 05:32:03 $
########################################################################
set license {
Copyright (c) 2002, Edwin A. Suominen
Registered Patent Agent * http://eepatents.com
Author, PRIVARIA Secure Networking Suite
Freely available at http://www.privaria.org

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

* Neither the name of the author nor any contributor may be used to
endorse or promote products derived from this software without
specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
}

# Load packages
foreach {i} {parse htmlgen} { package require $i }
# Load html commands into our namespace
namespace import -force ::htmlgen::*


#<<<<<<<<<<<< NAMESPACE ::webutils:: >>>>>>>>>>>>>>#
namespace eval ::webutils {

    namespace export -clear webgen tmlparse linksub tagsub keysub punctsub

}
#-----------------------------------------------


#<<<<<<<<<<<< ::webutils::webgen >>>>>>>>>>>>>>#
# webgen ?-file outPath? arrayName pageName
# outPath       = Path for .html page(s) to be written (created if not present)
# Returns HTML output for specified page, optionally also writes to outPath.html
proc ::webutils::webgen { args } {
    parse::args {file.arg} {arrayName pageName} $args

    # Setup
    set html {}         ;# HTML string that we build from null
    set sortList {
        ## TODO
    }

    # Start with any common elements, then go on to page-specific elements
    foreach {i} [list common $pageName] {

        foreach {j k} [array get $arrayName $i*] {

            # Process each element according to its type
            switch -glob $j {
                # Start with <HEAD>...</HEAD>
                ## TODO

            }
        }
    }

    return $html
}


#<<<<<<<<<<<< ::webutils::tmlparse >>>>>>>>>>>>>>#
# tmlparse arrayName specFile ?specFile2? ... ?specFileN?
# Returns list of page names defined by TML data
#
# TML File format:
# <<<???.styles>>>
# ...
# <<<???.domains>>>
# com gov edu de au etc. etc.
# ...
# <<<???.links>>>
# keyphrase linkexp
# ...
# <<<???.spans>>>
# keyphrase {attr value ...}
# ...
# <<<common.head>>>
# paramA {attr value pairs} {tag body} 
# ...
# paramB {attr value pairs} {tag body}
# ...
# <<<common.body>>>
# Plain text with [TCL procs] to generate any common HTML (at page top)
# ...
# <<<page1.tml>>>
# (or)
# <<<page1.head>>>
# paramC {attr value pairs} {tag body} 
# ...
# paramD {attr value pairs} {tag body}
# ...
# <<<page1.body>>>
# Plain text with [TCL procs] to generate any page1 HTML
# ...
proc ::webutils::tmlparse { args } {
    # Destination array name is first arg
    upvar [lindex $args 0] KEYS

    ### Read spec files into one big string
    set tmlString {}
    foreach {i} [lrange $args 1 end] {
        # Open settings file for reading
        set err [catch {set fh [open $i r]} errorMsg] 
        if { $err } {
            error "TML File Read Error: $errorMsg"

        } else {
            # Append each non-commented, non-blank line to string
            while { ![eof $fh] } {
                # Get first (or next) line of file
                gets $fh inLine
                # Append line if not blank or comment
                if { ![regexp {^#} $inLine] && ( [string length $inLine] > 1 ) } {
                    set tmlString "$tmlString\n$inLine"
                }
            }
            # Done reading file, so close it
            close $fh
        }
    }

    ### Iteratively substitute in all <<<xxx.tml>>> macros
    while {
        [regexp -nocase -indices {(<)<<[^>]+\.tml>>(>)} $tmlString null x y]
    } {
        set k1 [lindex $x 0]
        set k2 [lindex $y 0]
        set fname [string range "$tmlString" \
            [expr { $k1+3 }] [expr { $k2-3 }] ]
        if { ![catch { set fh [open $fname] }] } {
            set tmlString [string replace $tmlString $k1 $k2 [read $fh]]
            close $fh
        }
    }

    ### Get indices to all keywords
    set match 1; set k -1
    while {$match} {
        set match [regexp -start [expr $k+1] -indices \
            {(<)<<[^>]*>>(>)} $tmlString null x y]
        if {$match} {
             lappend keyList [lindex $x 0]
             set k [expr { [lindex $y 0]-1 }]
        }
    }

    ### Populate array with keyword sections, using keywords as indices
    # Create list of page names in the process
    set pageList {}
    set n [expr { [llength $keyList]-1 }]
    for {set i 0} {$i<$n} {incr i} {
        set k0 [lindex $keyList $i]
        set k1 [expr { [lindex $keyList [expr {$i+1}] ]-1 }]
        if { [regexp {<<<([^\.>]+)\.([^>]+)>>>(.*)} [string range $tmlString $k0 $k1] null x y z] } {
            array set KEYS [list $x.$y $z]
            # Add page name to list if not 'common' and not in list already
            if { ![string match $x common] && ( [lsearch $pageList $x] < 0 ) } {
                lappend pageList $x
            }
        }
    }

    return $pageList
}


#<<<<<<<<<<<< ::webutils::xmlsub (internal) >>>>>>>>>>>>>>#
proc ::webutils::xmlsub { tag attrList keyword string } {
    foreach {i j} $attrList {
        lappend attrs $i=$j
    }
    set x [eval ::xmlgen::doTag $tag $attrs $keyword]
    # Now do the full substitution
    regsub -all [subst {\x20$keyword\(\[^a-zA-Z<\])}] $string [subst {\x20$x\\1}] string

    return $string
}


#<<<<<<<<<<<< ::webutils::expsub (internal) >>>>>>>>>>>>>>#
proc ::webutils::expsub { exp string buffer subPlace } {
    # One loop iteration for each keyphrase match
    while {
        [regexp -indices "\[^a-zA-Z\<\>\]($exp)\[^a-zA-Z\<\>\]" $string null x]
    } {
        # Get match string
        set y [eval string range [subst -nocommands {{$string}}] $x]
        # In buffer, substitute match string for substitution
        # placeholder (e.g., "%1%")
        regsub -all $subPlace $buffer $y buffer
        # Replace match in main string with prepared buffer
        set string [eval string replace \
            [subst -nocommands {{$string}}] $x [subst {{$buffer}}] ]
    }
    return $string
}


#<<<<<<<<<<<< ::webutils::linksub >>>>>>>>>>>>>>#
# linksub ?-class <class>? ?-wwwclass <wwwclass>? ?-domains <list>? ?-autowww? subList string
# class         = CSS class name for normal links
# wwwclass      = CSS class name for links starting with "www."
# domains       = list of recognized domains, without leading "."
# subList       = keyword/link pairs. "http://" will be added if
#                 omitted and link is to a known domain from optional
#                 list -domains <list>
# Returns modified string
proc ::webutils::linksub { args } {
    parse::args {class.arg wwwclass.arg domains.arg autowww} {subList string} $args

    if { [info exists domains] } {
        set test "\{("
        foreach {i} $domains {
            set test "$test$i|"
        }
        set test "[string range $test 0 end-1])\}"
    } else {
        set test {impossible_regexp_match}
    }

    if { [info exists autowww] } {
        if { [info exists wwwclass] } {
            set moreAttr [list class $wwwclass]
        } else {
            set moreAttr {}
        }
        while { [regexp {[^>/]www\.[^\x20,]+} $string match] } {
            set string [xmlsub h [concat href $x $moreAttr] $match $string]
        }
    }

    foreach {i j} $subList {
        # Apply appropriate CSS class if specified
        if { [string match www.* $i] && [info exists wwwclass] } {
            set moreAttr [list class $wwwclass]
        } elseif { [info exists class] } {
            set moreAttr [list class $class]
        } else {
            set moreAttr {}
        }
        # Create the substitution string
        # No clue why the 'eval' is needed, but it is
        if { ![string match http://* $j] && [eval regexp $test $j] } {
        set x "http://$j"
        } else {
            set x $j
        }
        # Now do the full substitution for keyword in question
        set string [xmlsub h [concat href $x $moreAttr] $i $string]
    }

    return $string
}     


#<<<<<<<<<<<< ::webutils::keysub >>>>>>>>>>>>>>#
# keysub subList string
# subList       = { keyphrase {expanded result} ...}
# Returns modified string
proc ::webutils::keysub { args } {
    parse::args {} {subList string} $args
    
    foreach {i j} $subList {
        # Define what is substituted in, with %1% for enclosing target text
        buffer x $j
        # Now substitute each match to keyphrase with buffer contents,
        # except with any %1% in buffer replaced by match
        set string [expsub $i $string $x %1%]
    }

    return $string
}


#<<<<<<<<<<<< ::webutils::tagsub >>>>>>>>>>>>>>#
# tagsub subList stringName
# subList       = { keyTag1 keyTag2 {HTML %1%} ...}
# Modifies string in stringName
proc ::webutils::tagsub { args } {
    parse::args {} {subList stringName} $args

    foreach {i j k} $subList {
        # Define what is substituted in
        switch $j {
            ! {
                buffer x $k
                regsub %1% $x $i x
            } - {
                set x $k
            }
        }
        # Now do the full substitution
        # TODO
        regsub -all \
            [subst {(^|\\n)$i\\x20+(\[^\\n\]+)(\\n|$)}] \
            $string [subst {\\1$x\\2\\3}] string
    }

    return
}


#<<<<<<<<<<<< ::webutils::punctsub >>>>>>>>>>>>>>#
# punctsub ?-smart? stringName
# smart         = specify if proc should try using smart quotes
# Returns modified string
proc ::webutils::punctsub { args } {
    parse::args {smart} {string} $args

    set subList {
        {\ '(\[^\\x20\]|\[^s\]\[^\x20\])}       { \\&\#8216\;\\1}   { '\\1}
        {'}                                     {\\&\#8217\;}       {'}
        {(\[\\x20\(\])\"}                       {\\1\\&\#8220\;}    { \\&\#34\;\\1}
        {\"}                                    {\\&\#8221\;}       {\\&\#34\;}
        {&}                                     {\\&amp\;}          {\\&amp\;}
        {<}                                     {\\&lt\;}           {\\&lt\;}
        {>}                                     {\\&gt\;}           {\\&gt\;}
    }
    # Now do the substitutions for each punctuation mark found
    foreach {i j k} $subList {
        set x [expr { [info exists smart] ? $j : $k }]
        set keepLooping [regsub -all [subst $i] $string [subst $x] string]
    }

    return $string
}


#<<<<<<<<<<<< ::webutils::headset >>>>>>>>>>>>>>#
# headset ?-style <styleList>? headList stringName
# smart         = specify if proc should try using smart quotes
# Returns <style>...</style> block
proc ::webutils::headset { styleList } {

    # TODO
    # Proc body here

    return
}


#<<<<<<<<<<<< ::webutils::styleset (internal) >>>>>>>>>>>>>>#
# styleset styleList
# Returns <style>...</style> block
proc ::webutils::styleset { styleList } {

    # TODO
    # Proc body here

    return $result
}

